zjit/src/hir.rs

```diff
@@ -2270,24 +2323,32 @@ impl Function {
                         self.try_rewrite_uminus(block, insn_id, recv, state),
                     Insn::SendWithoutBlock { mut recv, cd, args, state, .. } => {
                         let frame_state = self.frame_state(state);
-                        let (klass, profiled_type) = if let Some(klass) = self.type_of(recv).runtime_exact_ruby_class() {
-                            // If we know the class statically, use it to fold the lookup at compile-time.
-                            (klass, None)
-                        } else {
-                            // If we know that self is reasonably monomorphic from profile information, guard and use it to fold the lookup at compile-time.
-                            // TODO(max): Figure out how to handle top self?
-                            let Some(recv_type) = self.profiled_type_of_at(recv, frame_state.insn_idx) else {
+                        let (klass, profiled_type) = match self.resolve_receiver_type(recv, self.type_of(recv), frame_state.insn_idx) {
+                            ReceiverTypeResolution::StaticallyKnown { class } => (class, None),
+                            ReceiverTypeResolution::Monomorphic { class, profiled_type }
+                            | ReceiverTypeResolution::SkewedPolymorphic { class, profiled_type }
+                            | ReceiverTypeResolution::SkewedMegamorphic { class, profiled_type } => (class, Some(profiled_type)),
```

Comment:
oh let's not use skewedmegamorphic --- we should fall back to SendWithoutBlockMegamorphic. this is a change from current behavior